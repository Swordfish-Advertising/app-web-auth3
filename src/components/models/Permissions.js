// @flow

type Permission = {
  streamId: string,
  level: 'read'|'contribute'|'manage',
  defaultName: ?string,
};

export type PermissionsList = Array<Permission>;

class Permissions {
  list: PermissionsList;

  constructor (permissionsList: string|PermissionsList) {
    if (typeof permissionsList === 'string') {
      this.list = JSON.parse(permissionsList);
    } else {
      this.list = permissionsList;
    }
  }

  updateList (newList: PermissionsList): PermissionsList {
    this.list = newList;
    return this.list;
  }
}

export default Permissions;
