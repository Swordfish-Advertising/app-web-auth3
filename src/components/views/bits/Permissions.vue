<template>
  <v-form>
    <h1>Permissions</h1>

    <div id="appIdText">
      App <b>{{ appId }}</b> is requesting:
    </div>

    <ul>
      <li
        v-for="(permission, index) in permissionsList"
        :key="index">
        A permission on stream <b>{{ permission.streamId }}</b> with level <b>{{ permission.level }}</b>
      </li>
    </ul>

    <div v-if="clientData.consentMsg">
      {{ clientData.consentMsg }}
    </div>

    <v-btn
      id="refusePermissions"
      @click="()=>this.$emit('refused')"
    >Reject</v-btn>

    <v-btn
      id="acceptPermissions"
      @click="()=>this.$emit('accepted')"
    >Accept</v-btn>

  </v-form>
</template>

<script>
export default {
  props: {
    permissionsList: {type: Array, default: () => ([])},
    appId: {type: String, default: ''},
    clientData: {type: Object, default: () => ({})},
  },
};
</script>
