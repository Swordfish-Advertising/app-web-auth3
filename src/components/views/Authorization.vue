<template>
  <div>
    <Permissions v-if="permissionsArray && appId" :appId="appId" :permissionsArray="permissionsArray"></Permissions>

    <v-form ref="form" v-model="validForm" v-else>
      <h1>Sign in</h1>

      <v-text-field
        label="Username or email"
        :rules="[rules.required]"
      ></v-text-field>

      <Password></Password>

      <v-btn
        id="submitButton"
        @click="submit"
        :disabled="!validForm"
      >Sign In</v-btn>

      <v-btn
      >Cancel</v-btn>
      
      <NavigationButton
        :title="'Create an account'"
        :page="'register'"
      ></NavigationButton>

      <NavigationButton
        :title="'Forgot password'"
        :page="'reset'"
      ></NavigationButton>

      <div>
        Feel free to reach our
        <a target="_blank" href="https://pryv.com/helpdesk">helpdesk</a>
        if you have questions.
      </div>
    </v-form>
  </div>
</template>

<script>
  import Password from './bits/Password.vue';
  import NavigationButton from './bits/NavigationButton.vue';
  import Permissions from './bits/Permissions.vue';

  export default {
    components: {
      Password,
      NavigationButton,
      Permissions
    },
    props: ['permissionsArray', 'appId'],
    data: () => ({
      rules: {
        required: value => !!value || 'This field is required.',
        email: value => /.+@.+/.test(value) || 'E-mail must be valid'
      },
      validForm: false
    }),
    methods: {
      submit () {
        if (this.$refs.form.validate()) {
          alert('Submit form...')
        }
      }
    }
  }
</script>