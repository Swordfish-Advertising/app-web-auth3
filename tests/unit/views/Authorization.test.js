import { shallowMount } from '@vue/test-utils';
import Authorization from '@/components/views/Authorization';
import Permissions from '@/components/views/bits/Permissions';
import Vue from 'vue';
import Vuetify from 'vuetify';
import VueRouter from 'vue-router';

Vue.use(Vuetify);
Vue.use(VueRouter);

describe('Authorization.test.js', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(Authorization);
  });

  it('renders correctly (snapshots matching)', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('deactivates submit button when form is not yet filled', () => {
    const submitButton = wrapper.find('#submitButton');
    expect(wrapper.vm.validForm).toBe(false);
    expect(submitButton.attributes().disabled).toBe('true');
  });

  it('activates submit button when form is valid', () => {
    wrapper.setData({
      validForm: true,
    });
    const submitButton = wrapper.find('#submitButton');
    expect(submitButton.attributes().disabled).toBeFalsy();
  });

  it('validates required fields correctly', () => {
    const requiredRule = wrapper.vm.rules.required;
    expect(requiredRule()).toBe('This field is required.');
    expect(requiredRule('notEmpty')).toBe(true);
  });

  it('shows authorization form by default', () => {
    // Renders auth form
    const authForm = wrapper.find({ref: 'form'});
    expect(authForm.exists()).toBe(true);
    // Does not render permissions
    const permissions = wrapper.find(Permissions);
    expect(permissions.exists()).toBe(false);
  });

  it('shows requested permissions when app access check is successful', () => {
    wrapper.setData({
      permissionsList: [{streamId: 'diary', level: 'read'}],
    });
    // Does not render auth form
    const authForm = wrapper.find({ref: 'form'});
    expect(authForm.exists()).toBe(false);
    // Renders permissions instead
    const permissions = wrapper.find(Permissions);
    expect(permissions.exists()).toBe(true);
  });
});
