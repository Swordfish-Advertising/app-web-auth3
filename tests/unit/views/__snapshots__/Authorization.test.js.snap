// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authorization.test.js renders correctly (snapshots matching) 1`] = `
<div>
  <valert-stub
    transition="scale-transition"
    type="error"
    value="{}"
  >
    {}
  </valert-stub>
   
  <vform-stub>
    <h1>
      Sign in
    </h1>
     
    <vtextfield-stub
      id="usernameOrEmail"
      label="Username or email"
      rules="function required(value) {
          return !!value || 'This field is required.';
        }"
    />
     
    <password-stub
      value=""
    />
     
    <vbtn-stub
      activeclass="v-btn--active"
      disabled="true"
      id="submitButton"
      ripple="true"
      tag="button"
      type="button"
    >
      Sign In
    </vbtn-stub>
     
    <vbtn-stub
      activeclass="v-btn--active"
      ripple="true"
      tag="button"
      type="button"
    >
      Cancel
    </vbtn-stub>
     
    <!---->
     
    <router-link-stub
      event="click"
      tag="a"
      to="[object Object]"
    >
      Create an account
    </router-link-stub>
     
    <router-link-stub
      event="click"
      tag="a"
      to="[object Object]"
    >
      Forgot password
    </router-link-stub>
  </vform-stub>
</div>
`;
